Fix css-flexbox/flex-aspect-ratio-img-row-015

sammygill opened the pull request:
#### 9944033dacc36097f372ba5497c5b245ed8cdf88
<pre>
Fix css-flexbox/flex-aspect-ratio-img-row-015
<a href="https://rdar.apple.com/84995186">rdar://84995186</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=232408">https://bugs.webkit.org/show_bug.cgi?id=232408</a>

Reviewed by Alan Baradlay.

In the WPT, we have a row flexbox with a single flex item which is an SVG
which does not have an intrinsic size but has an intrinsic aspect ratio.
When computing the base size of this flex item, we compute its preferred
logical width.

Since this SVG does not have any intrinsic sizes, we end up computing the
stretch fit. When this occurs, we also need to consider any min/max
constraints imposed upon the item from the opposite axis. This ends up
constraining the logical width, and as a result, the base size, instead
of it being as wide as its containing block. We can achieve this by
using computeMinMaxLogicalWidthFromAspectRatio which computes these
transferred constraints and then clamping the stretch fit size between
these values.

Canonical link: <a href="https://commits.webkit.org/298651@main">https://commits.webkit.org/298651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b4b601cbca0e2f3c78b509418aea920d5f927be

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26334 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122184 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66684 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0667f29e-865c-4f0f-bac6-d9a85f0a4d19) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44377 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88230 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9791af2e-dbfb-4c77-8b04-90c9ebf76d53) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119076 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29105 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104204 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68640 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28233 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22313 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65866 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98496 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22448 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125334 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43022 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96960 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43387 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100393 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96744 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24626 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42019 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19907 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38969 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48501 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42376 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45711 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44080 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -708,7 +709,10 @@ LayoutUnit RenderReplaced::computeReplacedLogicalWidth(ShouldComputePreferred sh
+
                 auto constrainedLogicalWidth = computeConstrainedLogicalWidth();
+                auto [transferredMinLogicalWidth, transferredMaxLogicalWidth] = computeMinMaxLogicalWidthFromAspectRatio();
+                constrainedLogicalWidth = std::clamp(constrainedLogicalWidth, transferredMinLogicalWidth, transferredMaxLogicalWidth);
```
I hope transferredMinLogicalWidth is never greater than transferredMaxLogicalWidth as std::clamp gets really confused in such cases.

sammygill replied:
Seems like this should be ok currently since `computeMinMaxLogicalWidthFromAspectRatio()` makes sure that is true before returning... but I also added a debug assert to help catch this if there are any potential changes to that function

sammygill committed:
Fix css-flexbox/flex-aspect-ratio-img-row-015
rdar://84995186
https://bugs.webkit.org/show_bug.cgi?id=232408

Reviewed by Alan Baradlay.

In the WPT, we have a row flexbox with a single flex item which is an SVG
which does not have an intrinsic size but has an intrinsic aspect ratio.
When computing the base size of this flex item, we compute its preferred
logical width.

Since this SVG does not have any intrinsic sizes, we end up computing the
stretch fit. When this occurs, we also need to consider any min/max
constraints imposed upon the item from the opposite axis. This ends up
constraining the logical width, and as a result, the base size, instead
of it being as wide as its containing block. We can achieve this by
using computeMinMaxLogicalWidthFromAspectRatio which computes these
transferred constraints and then clamping the stretch fit size between
these values.

Canonical link: https://commits.webkit.org/298651@main
